lib/xpedite/framework/session/LocalSession.H

```diff
@@ -37,27 +38,37 @@ namespace xpedite { namespace framework { namespace session {
     public:
 
     LocalSession(Handler& handler_)
-      : _request {}, _handler (handler_), _isAlive {} {
+      : _requestEscrow {}, _handler (handler_), _isAlive {} {
     }
 
-    bool execute(request::Request* request_, MilliSeconds timeout_) {
+    template <typename RequestType>
+    bool execute(RequestType* request_, MilliSeconds timeout_) {
       MilliSeconds elapsed {0};
+      auto request = std::make_unique<RequestType>(*request_);
       // enque request
       while(!hasTimedOut(timeout_, elapsed)) {
-        request::Request* expected = {};
-        if(_request.compare_exchange_weak(expected, request_, std::memory_order_release, std::memory_order_relaxed)) {
-          break;
-        }
+	if (_requestEscrow.tryDeposit(request.get())) {
+	  break;
+	}
+
         std::this_thread::sleep_for(_handler.pollInterval());
         elapsed += _handler.pollInterval();
+	break;
       }
       
       // await execution
-      while(!hasTimedOut(timeout_, elapsed) && _request.load(std::memory_order_acquire) == request_) {
+      while(!hasTimedOut(timeout_, elapsed) && !_requestEscrow.empty()) { 
         std::this_thread::sleep_for(_handler.pollInterval());
         elapsed += _handler.pollInterval();
       }
 
+      const auto releasedPointer = _requestEscrow.releaseOwnership();
+      if (!releasedPointer) {
+	*request_ = std::move(*request);
```

Comment:
request::Request* is a base type. This move may not invoke the move/copy constructor for children.
we could just copy/move `request->response()` to `request_->response()`